### PR TITLE
implement box zoom

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request implements box and x-range zoom capability across all applicable viewers.

* **imviz**: box and box-match
* **specviz**: box
* **cubeviz**: box for both image and spectral viewers
* **mosviz**: box for image and 1d spectral viewers, xrange and box for 2d spectral viewer (with vertical pan *removed* - see screenshot in comments below, recording is out-dated slightly)

Notes: 
* This also renames all tools created in jdaviz to remove "bqplot" from the names in order to avoid future confusion.  If this causes backwards compatibility issues or concerns, it can be reverted or handled as a separate PR.
* In Imviz, the additional two buttons begin to cause space overflow issues with two side-by-side viewers in the default width (see in recording below).  Down the road, we may want to consider "dropdowns" for the matched zoom options, or a separate "zoom lock" button at the app level similar to mosviz.
* When equal aspect ratio is turned on, the resulting zoom will contain the entire selected rectangle (but in doing so, may include more in one dimension than selected, see the end of the imviz recording below).

https://user-images.githubusercontent.com/877591/147145208-d454c6f0-7f89-4bd7-837e-1ca1af0af2bd.mov

https://user-images.githubusercontent.com/877591/147145216-bc5b49db-ada4-40eb-a50e-f8635f90d32d.mov

**TODO**:
- [x] replace placeholder icons with custom icons from Jenn.  Preferably after #995 is merged.
- [x] ~better handling of equal aspect ratio support~ (defer for feedback: one alternate option may be to restrict the rectangle to the aspect of the current viewer, but that will be difficult to implement and may just add confusion)
- [x] changes entry
- [x] screen recording(s) in PR description

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:25]: Are you sure this is not a new feature?

File: jdaviz/configs/imviz/plugins/tools.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/tools.py:140]: I don't see how this is useful in Imviz. Can you please clarify?

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/tools.py:140]: this is used in the 2d spectrum viewer in mosviz (was just as easy to create it now while making sure the flexibility exists)

Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py:140]: I am just confused why it is hidden inside the `imviz` directory.

File: CHANGES.rst
Comment (kecnry) [reply CHANGES.rst:25]: woops, good catch.  Should this go at the top of "New Features" (since it is applicable to multiple configurations/viewers) or in "Other Changes and Additions"?

File: jdaviz/configs/imviz/plugins/tools.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/tools.py:140]: ok, let me try to move all that are not imviz-specific into some generic location higher up (there does not yet seem to be a place for custom general tools, so I guess its time to make one).

Comment (pllim) [reply jdaviz/configs/imviz/plugins/tools.py:140]: `jdaviz/core` or `jdaviz/configs/default`? 🤷

Comment (kecnry) [reply jdaviz/configs/imviz/plugins/tools.py]: done (moved all non-imviz to `jdaviz.core.tools` and left everything imviz-specific behind)

File: CHANGES.rst
Comment (kecnry) [reply CHANGES.rst]: moved to "Other Changes and Additions" - if you think it belongs better elsewhere, let me know!

Comment (pllim) [reply CHANGES.rst]: IMHO "new features" is better suited because this is a nice cool new feature but I'll leave it to your good judgement. Either way, since it affects multiple layouts, it can be outside the subsections.

**Review by pllim — APPROVED** [review]
Cubeviz: Works fine, though it made me realized how nice to have "locked pan/zoom" across the viewers like Imviz (future work!).

Imviz: Works great. Even in dark mode! 😉 

Mosviz:
* On the image viewer, with lock active, the zoom is not preserved when I switch images.  I am not sure if fixing this is in scope or not.
* Maybe future work: Would be nice to not limit 2D spec zoom to X-range. One might just want to average out a narrower area in the Y-direction too to reduce background noise in the collapsed 1D spectrum.

Specviz: Works fine.

File: jdaviz/configs/imviz/plugins/tools.py
Comment (pllim) [inline jdaviz/configs/imviz/plugins/tools.py:22]: Is it hard for me to see from diff but you might want to make sure that we delete icons that are no longer necessary.

**Comment by pllim** [comment]
> backwards compatibility

I don't think it is an issue. Icon locations are not something that users care about.

> space overflow

Agreed. Hopefully some of these will naturally be resolved when we implement the wireframe designs.

> equal aspect ratio

Is this a common use case? If not, not high priority to fix.

File: jdaviz/configs/imviz/plugins/tools.py
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/tools.py:22]: these icons are still used, the code to modify them was just moved from `imviz.plugin.tools` to `jdaviz.core.tools`.  The only unused icons currently in the repo are a yrange zoom (currently unused but I requested the icon since we may need it in the future) and an additional pan icon.

**Comment by pllim** [comment]
No tests added and yet coverage increased. Hmm! 😉 

Also, any user-facing documentation needs updating here? Example: https://jdaviz.readthedocs.io/en/latest/specviz/displaying.html#pan-zoom

**Comment by kecnry** [comment]
@pllim - l would consider the locking issue in scope, I'll look into that and fix it before we merge this.  I can also easily add a box zoom to the mosviz 2d spectrum viewer (there already is a 2d pan/zoom, I was just trying to avoid adding too many options).  Giving that a little more thought, I think a box zoom would _replace_ the need for the vertical pan/zoom - so unless there are any objections, I'll go ahead with that and can always tweak with any feedback we get.

**Comment by pllim** [comment]
@kecnry , if you are in doubt, ask a PO first. I wouldn't want you to spend time on it only to be told to undo your changes... 😬

**Comment by kecnry** [comment]
Edit: row locking does still work as designed but is limited to the **x-range** on the 1d and 2d spectrum viewers as well as the stretch/percentile options on the image viewer and 2d spectrum viewers (see #918).  I don't think it makes sense to lock the image limits, but we may want to reconsider the y-limits on the 2d spectrum viewer in the future. 

Mosviz 2d image viewer toolbar now has the following options:
Screen Shot 2021-12-22 at 6 01 41 PM

**Review by rosteen — APPROVED** [review]
Looks great! I've been wanting this for a long time...tested with all the configs and was unable to break it.

**Comment by pllim** [comment]
> unable to break it

**Pull request merged**
